File: app/src/test/java/com/moviereel/ui/activities/intro/SplashPresenterTest.java
package com.moviereel.ui.activities.intro;

import com.moviereel.data.DataManager;
import com.moviereel.ui.intro.splash.SplashPresenterImpl;
import com.moviereel.ui.intro.splash.SplashView;
import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.junit.MockitoJUnitRunner;
import java.util.concurrent.Callable;
import io.reactivex.Scheduler;
import io.reactivex.android.plugins.RxAndroidPlugins;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

import static org.mockito.Mockito.verify;


/**
 * @author lusinabrian
 */
@RunWith(MockitoJUnitRunner.class)
public class SplashPresenterTest {

    @Mock
    SplashView mMockSplashView;

    @Mock
    DataManager mMockDataManager;

    private SplashPresenterImpl<SplashView> mSplashPresenterImpl;

    @BeforeClass
    public static void onlyOnce() throws Exception{
        RxAndroidPlugins.setInitMainThreadSchedulerHandler(new Function<Callable<Scheduler>, Scheduler>() {
            @Override
            public Scheduler apply(@NonNull Callable<Scheduler> schedulerCallable) throws Exception {
                return Schedulers.trampoline();
            }
        });
    }

    @Before
    public  void setUp() throws Exception{
        MockitoAnnotations.initMocks(this);
        CompositeDisposable compositeDisposable = new CompositeDisposable();
        mSplashPresenterImpl = new SplashPresenterImpl<>(mMockDataManager, compositeDisposable);
        mSplashPresenterImpl.onAttach(mMockSplashView);
    }

    @After
    public void tearDown() throws Exception{
        RxAndroidPlugins.reset();
        mSplashPresenterImpl.onDetach();
    }

    @Test
    public void testSplashScreenShouldOpenIntroActivityOnFirstStart(){
        verify(mMockSplashView).openMainActivity();
    }

    @Test
    public void testSplashScreenShouldOpenMainActivityOnSecondStart(){
        mMockSplashView.openAppIntroductionActivity();

        verify(mMockSplashView).openMainActivity();
    }

    @Test
    public void testSplashScreenStartsSyncService(){
        verify(mMockSplashView).startSyncService();
    }
}

File: mobile-ui/src/test/kotlin/com/moviereel/ui/intro/SplashPresenterTest.kt
package com.moviereel.ui.intro

import com.moviereel.TestSchedulerProvider
import com.moviereel.data.DataManager
import com.moviereel.ui.intro.splash.SplashPresenter
import com.moviereel.ui.intro.splash.SplashPresenterImpl
import com.moviereel.ui.intro.splash.SplashView
import io.reactivex.android.plugins.RxAndroidPlugins
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.schedulers.TestScheduler
import org.junit.After
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.mockito.junit.MockitoJUnitRunner


/**
 * @author lusinabrian
 */
@RunWith(MockitoJUnitRunner::class)
class SplashPresenterTest {

    @Mock lateinit var mMockSplashView: SplashView
    @Mock lateinit var mMockDataManager: DataManager
    lateinit var mTestScheduler: TestScheduler

    lateinit var mSplashPresenter: SplashPresenter<SplashView>

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        val compositeDisposable = CompositeDisposable()
        mTestScheduler = TestScheduler()
        val testSchedulerProvider = TestSchedulerProvider(mTestScheduler)

        mSplashPresenter = SplashPresenterImpl(mMockDataManager, testSchedulerProvider, compositeDisposable)
        mSplashPresenter.onAttach(mMockSplashView)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        RxAndroidPlugins.reset()
        mSplashPresenter.onDetach()
    }

    @Test
    fun testSplashScreenShouldOpenIntroActivityOnFirstStart() {
        verify<SplashView>(mMockSplashView).openMainActivity()
    }

    @Test
    fun testSplashScreenShouldOpenMainActivityOnSecondStart() {
        mMockSplashView.openAppIntroductionActivity()

        verify<SplashView>(mMockSplashView).openMainActivity()
    }

    @Test
    fun testSplashScreenStartsSyncService() {
        verify<SplashView>(mMockSplashView).startSyncService()
    }

    companion object {

        @BeforeClass
        @Throws(Exception::class)
        fun onlyOnce() {
            RxAndroidPlugins.setInitMainThreadSchedulerHandler { Schedulers.trampoline() }
        }
    }
}